src/main/java/stirling/software/SPDF/model/provider/Provider.java

```diff
@@ -0,0 +1,133 @@
+package stirling.software.SPDF.model.provider;
+
+import static stirling.software.SPDF.model.UsernameAttribute.EMAIL;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.stream.Collectors;
+
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import stirling.software.SPDF.model.UsernameAttribute;
+import stirling.software.SPDF.model.exception.UnsupportedUsernameAttribute;
+
+@Data
+@NoArgsConstructor
+public class Provider {
+
+    public static final String EXCEPTION_MESSAGE = "The attribute %s is not supported for %s.";
+
+    private String issuer;
+    private String name;
+    private String clientName;
+    private String clientId;
+    private String clientSecret;
+    private Collection<String> scopes;
+    private UsernameAttribute useAsUsername;
+    private String logoutUrl;
+    private String authorizationUri;
+    private String tokenUri;
+    private String userInfoUri;
+
+    public Provider(
+            String issuer,
+            String name,
+            String clientName,
+            String clientId,
+            String clientSecret,
+            Collection<String> scopes,
+            UsernameAttribute useAsUsername,
+            String logoutUrl,
+            String authorizationUri,
+            String tokenUri,
+            String userInfoUri) {
+        this.issuer = issuer;
+        this.name = name;
+        this.clientName = clientName;
+        this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        this.scopes = scopes == null ? new ArrayList<>() : scopes;
+        this.useAsUsername =
+                useAsUsername != null ? validateUsernameAttribute(useAsUsername) : EMAIL;
+        this.logoutUrl = logoutUrl;
+        this.authorizationUri = authorizationUri;
+        this.tokenUri = tokenUri;
+        this.userInfoUri = userInfoUri;
+    }
+
+    public void setScopes(String scopes) {
+        if (scopes != null && !scopes.isBlank()) {
+            this.scopes =
+                    Arrays.stream(scopes.split(",")).map(String::trim).collect(Collectors.toList());
+        }
+    }
+
+    private UsernameAttribute validateUsernameAttribute(UsernameAttribute usernameAttribute) {
+        switch (name) {
+            case "google" -> {
+                return validateGoogleUsernameAttribute(usernameAttribute);
+            }
+            case "github" -> {
+                return validateGitHubUsernameAttribute(usernameAttribute);
+            }
+            case "keycloak" -> {
+                return validateKeycloakUsernameAttribute(usernameAttribute);
+            }
+            default -> {
+                return usernameAttribute;
+            }
+        }
+    }
+
+    private UsernameAttribute validateKeycloakUsernameAttribute(
+            UsernameAttribute usernameAttribute) {
+        switch (usernameAttribute) {
+            case EMAIL, PREFERRED_NAME -> {
```

Comment:
I missing PREFERRED_USERNAME

`Caused by: stirling.software.SPDF.model.exception.UnsupportedUsernameAttribute: The attribute PREFERRED_USERNAME is not supported for Keycloak.`